store/tikv/region_cache.go

```diff
@@ -604,6 +654,13 @@ func (r *Region) Contains(key []byte) bool {
 		(bytes.Compare(key, r.meta.GetEndKey()) < 0 || len(r.meta.GetEndKey()) == 0)
 }
 
+// ContainsByEnd checks whether the key is in the region, for the maximum region endKey is empty.
+// startKey < key <= endKey.
+func (r *Region) ContainsByEnd(key []byte) bool {
+	return bytes.Compare(r.meta.GetStartKey(), key) < 0 &&
+		(bytes.Compare(key, r.meta.GetEndKey()) <= 0 || len(r.meta.GetStartKey()) == 0)
```

Comment:
What is the expected behavior when `key == nil`?

Reply:
That means the lower bound is nil. So false is expected.

Reply:
Why the comment of `ContainsByEnd` is exactly the same as `Contains`... I think `ContainsByEnd` will check if the region contains the greatest key that is less than `key`.
And in order to make it possible to reverse seek from the last region in tikv,  the argument `key` should be _EndKey_, which means if `nil` is passed in, it means _+Inf_.

Reply:
That's right. I will edit comments.

I thought nil was the key smaller than [] byte {0}. I will try to change used as the upper limit.

Reply:
It is somewhat tricky we handle region keys. Whether `nil` is less or greater than `\0` depends on the key is start key or end key. For region start key, `nil` means -Inf. For region end key, `nil` means +Inf.

Reply:
I agree to use reverse scan when scanning from the last item. However, it seems that we can not get the last region with the PD API.

Suggesting that it might be possible to implement PD API. However, I think this function is unnecessary because I think that there are not many when scanning from the end.

What do you think?

https://github.com/pingcap/kvproto/blob/54cf0a0dfe552852805379ddad4fd24a43fd10b9/proto/pdpb.proto#L40

https://github.com/pingcap/pd/blob/9d792fba552db0e8419bd2658ac54f44c0192a4d/server/core/region_tree.go#L123-L138

Reply:
OK. It makes sense. So we should specify that `reverse_seek(nil)` always get empty result.

Reply:
I already think that it is implemented in tikv like that.

Reply:
Emmmm... Should the key `""` contained in a region whose startKey and endKey are both `""`?

Reply:
In that case, searchCachedRegion will always return nil.
```go
if isEndKey && bytes.Compare(r.StartKey(), key) == 0 {
  r = nil     // clear result
  return true // iterate next item
}
```

Reply:
Emm... In my opinion, `isEndKey == true` means that the given key is an endKey of something, and in this context the key is empty means the very end of the whole range. So `""` is the endKey of region whose range is `["", "")`, which means, this function should return false.....